state-chain/pallets/cf-trading-strategy/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+									// checked the balance above
+									log_or_panic!(
+										"Failed to update limit order for strategy {strategy_id:?}"
+									);
+								}
+							}
+						};
+
+						if remaining_weight.checked_sub(&new_weight_estimate).is_none() {
+							break;
+						}
+
+						update_limit_order_from_balance(base_asset, Side::Sell, sell_tick);
+						update_limit_order_from_balance(STABLE_ASSET, Side::Buy, buy_tick);
+					},
+				}
+			}
+
+			weight_used
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		StrategyDeployed {
+			account_id: T::AccountId,
+			strategy_id: T::AccountId,
+			base_asset: Asset,
+			strategy: TradingStrategy,
+		},
+		FundsAddedToStrategy {
+			strategy_id: T::AccountId,
+			base_asset: Asset,
+			base_asset_amount: AssetAmount,
+			quote_asset_amount: AssetAmount,
+		},
+		StrategyClosed {
+			strategy_id: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		StrategyNotFound,
+		AmountBelowDeploymentThreshold,
+		InvalidOwner,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn deploy_trading_strategy(
+			origin: OriginFor<T>,
+			base_asset_amount: AssetAmount,
+			quote_asset_amount: AssetAmount,
+			base_asset: Asset,
+			strategy: TradingStrategy,
+		) -> DispatchResult {
+			let lp = &T::AccountRoleRegistry::ensure_liquidity_provider(origin)?;
+
+			let quote_asset = STABLE_ASSET;
+
+			let strategy_id = {
+				// Check that strategy is created with sufficient funds:
+				{
+					let fraction_of_required = |asset, provided| {
+						let min_required = MinimumDeploymentAmountForStrategy::<T>::get(asset);
+
+						if min_required == 0 {
+							FixedU64::one()
+						} else {
+							FixedU64::from_rational(provided, min_required)
+						}
+					};
+
+					ensure!(
+						fraction_of_required(base_asset, base_asset_amount) +
+							fraction_of_required(quote_asset, quote_asset_amount) >=
+							FixedU64::one(),
+						Error::<T>::AmountBelowDeploymentThreshold
+					);
+				}
+
+				let nonce = frame_system::Pallet::<T>::account_nonce(lp);
+
+				let strategy_id = derive_strategy_id::<T>(lp, nonce);
+
+				Self::deposit_event(Event::<T>::StrategyDeployed {
+					account_id: lp.clone(),
+					strategy_id: strategy_id.clone(),
+					base_asset,
+					strategy: strategy.clone(),
+				});
+
+				Strategies::<T>::insert(
+					strategy_id.clone(),
+					TradingStrategyEntry { base_asset, owner: lp.clone(), strategy },
+				);
+
+				strategy_id
+			};
+
+			Self::add_funds_to_existing_strategy(
+				lp,
+				&strategy_id,
+				base_asset,
+				base_asset_amount,
+				quote_asset_amount,
+			)
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn close_strategy(origin: OriginFor<T>, strategy_id: T::AccountId) -> DispatchResult {
```

Comment:
I believe there's a gap when cancelling orders:
1. Create an LP account
2. Set refund address for USDC
3. Deposit USDC
4. Enter USDC/USDT strategy, with just USDC
5. Trades filled
6. Cancel strategy
7. You now have USDT but no refund address for it.

We check the refund addresses in `try_mutate_order` when cancelling the orders, however, at this point, the `lp` is actually the strategy id which doesn't have refund addresses. 

Recommendation would be to check this when deploying a strategy, that the user has a refund address for both assets.

Also test ofc.

Reply:
I did consider refund addresses for strategies, and I figured a reasonable approach would be for all strategies to use refund addresses of their owners (and, as discussed, it wasn't deemed necessary for this PR since we don't have code for emergency refunding yet).

I did not consider `try_mutate_order`, however. What is the rationale for checking refund addresses there? I think it should be sufficient to check in extrinsics only, and if orders are updated internally (as is the case for automated strategies), this check can be omitted. Does that sound reasonable? Also would be interested in @dandanlen's opinion.

Reply:
I do agree that try_mutate_order seems to be a strange place to have those checks

Reply:
I've split the function, but in my opinion it is better to do as I sugested above and simply add these check at the extrinsics level. The new `inner_update_limit_order_with_refund_address_checks` is only used in one place -- the `update_limit_order` extrinsic -- and we could simplify code considerably if we just move the checks there. (Then we could do the same for `set_limit_order`: here I didn't bother creating another function for inner_set_limit_order without checks since it is not needed, but for consistency we couls also move refund address checks to the extrinsic).

Reply:
Struggling to parse the diff right now, my brain is a bit fried. Will look again later. 

Either way, it seems to work? I do agree putting checks in the extrinsic is generally better, but in this case there are multiple extrinsics that all end up calling into the same (originally private?) methods, so pushing the checks down into the inner update fn kind of made sense at the time. 

What I meant with my comment was that ideally there would have been a cleaner separation and the LPs would call into the LP pallet via a single extrinsic, the checks would happen there, and then the pools pallet just exposes the interface for adding/updating orders. Bit late for that now of course...

Reply:
My main point is that the new `inner_update_limit_order_with_refund_address_checks` is only called in one place, not multiple places. There wouldn't be any duplication if move the check and remove `inner_update_limit_order_with_refund_address_checks`.

Reply:
> My main point is that the new inner_update_limit_order_with_refund_address_checks is only called in one place, not multiple places. There wouldn't be any duplication if move the check and remove inner_update_limit_order_with_refund_address_checks.

Agree, I think having the checks where they are now in `try_mutate_order` is unintuitive, and isn't consistent with how we now do it for update order. Better to pull them out into set_order so it's consistent between the two extrinsics, remove the single-use function and then not having `try_mutate_order` do the unexpected check.

Reply:
Just discussed this IRL with Kyle - I think we're all in agreement and all agree with moving checks out into the extrinsics- shall we just to this? I can add a commit with this change and we can discuss if necessary.